### PR TITLE
add a flatten method

### PR DESCRIPTION
**Pull request by Announcement**
# usage
~~~ javascript
(function(){
    var a = Lazy([1, 2])
    var b = Lazy([3, 4])
    var c = Lazy([5, 6])
    var d = Lazy([a, b])
    var e = Lazy([c])
    var f = Lazy([d, e])
    var g = Lazy(f)
    var h = g.flatten();
  
    [a, b, c, d, e, f, g, h].forEach(it => console.log(it))
  }())
~~~

### CODE REVIEW

**Comment by Announcement** [comment]
related issue https://github.com/dtao/lazy.js/issues/125

File: lazy.js [hunk 1]
```diff
@@ -6628,6 +6628,37 @@
...
+
+    source = this.source
+    queue = []
+    results = []
+
+    source.forEach(it => queue.push(it))
+
+    while (queue.length) handle(queue.shift())
+
+    this.source = results
```
Comment (dtao) [inline lazy.js]: So this mutates the current sequence. This isn't how this library works. Take a look at any of the other methods for a better sense of how to implement this.

File: lazy.js [hunk 2]
```diff
@@ -6628,6 +6628,37 @@
     return snapshot;
   };
 
+  Sequence.prototype.flatten = function flatten () {
```
Comment (dtao) [inline lazy.js]: This method already exists. If the goal is to add a non-recursive `flatten` method per #125, I think the answer is to add support for a `depth` flag or something like that to `FlattenedSequence`, not to start from scratch.

File: lazy.js [hunk 3]
```diff
@@ -6628,6 +6628,37 @@
...
+    var results
+    var item
+
+    source = this.source
+    queue = []
+    results = []
+
+    source.forEach(it => queue.push(it))
+
+    while (queue.length) handle(queue.shift())
```
Comment (dtao) [inline lazy.js]: Nothing about this implementation seems lazy. :/

**Comment by dtao** [comment]
I appreciate the contribution! I left a few comments on the code. While I welcome pull requests in general, unfortunately I think the code as written right now doesn't fit in this library at all. With that in mind I'm going to close this, but please don't feel discouraged from trying again.

**Pull request closed without merging**